### PR TITLE
【MIIT program】add GPWNO model and MD/QM9/MP configs

### PR DESCRIPTION
**Pull request by yqf66**

### CODE REVIEW

**Comment by leeleolay** [comment]
https://github.com/PaddlePaddle/PaddleMaterials/issues/258
辛苦参考这个标准对PR做补充

**Comment by yqf66** [comment]
### PR 类型

模型复现 / 新模型接入

### 关联 Issue

- Contributing guide: https://github.com/PaddlePaddle/PaddleMaterials/issues/258
- Existing PR update: https://github.com/PaddlePaddle/PaddleMaterials/pull/263

### 本次提交文件

本 PR 仅提交代码、配置和文档，不提交数据集、checkpoint、log、tensorboard 或本地输出目录。

- `ppmat/models/gpwno/GPWNO.py`
- `ppmat/models/gpwno/GPWNO_utils.py`
- `ppmat/models/gpwno/PWNO_utils.py`
- `ppmat/models/__init__.py`
- `electronic_structure/configs/gpwno/README.md`
- `electronic_structure/configs/gpwno/gpwno_md17_benzene.yaml`
- `electronic_structure/configs/gpwno/gpwno_md17_ethane.yaml`
- `electronic_structure/configs/gpwno/gpwno_mp.yaml`
- `electronic_structure/README.md`

### 主要改动

- 新增 GPWNO 电子密度预测模型实现，接入 `ppmat.models` 的 `build_model` 流程。
- 新增 `electronic_structure/configs/gpwno/` 下 MD17_EC 与 MP_EC 训练配置。
- 新增 GPWNO README，包含模型简介、数据集说明、关键配置、训练/评测/推理命令和当前复现实验结果。
- 更新 `electronic_structure/README.md` 模型矩阵，加入 GPWNO。

### 当前复现结果

- `gpwno_md17_benzene`: validation NMAE `3.5567%`，10 epochs，本地 run `gpwno_md17_benzene_t_20260527_182931_s_42`。
- `gpwno_md17_ethane`: test NMAE `4.8339%`，本地 run `gpwno_md17_ethane_t_20260529_204928_s_42`。
- `gpwno_mp`: test NMAE `37.8910%`，本地 run `gpwno_mp_resume_t_20260528_204842_s_42`。

### Checkpoint / Dataset / Log 说明

- 本 PR 不提交 checkpoint、dataset、log、tensorboard 和任何本地输出目录。
- README 的 `Checkpoint | Log` 链接位置已保留为 `TBD`。
- 相关 checkpoint 将单独提供给工程方上传，等待工程方转换为 Paddle 专属链接后，再回填 README。


### 验证

- `python -m py_compile ppmat/models/gpwno/GPWNO.py ppmat/models/gpwno/GPWNO_utils.py ppmat/models/gpwno/PWNO_utils.py ppmat/models/__init__.py ppmat/datasets/density_dataset.py`
- 使用 OmegaConf 解析 GPWNO YAML 配置：`gpwno_md17_benzene.yaml`、`gpwno_md17_ethane.yaml`、`gpwno_mp.yaml`

### 说明

- 本 PR 沿用 `electronic_structure/train.py` / `predict.py` 入口和现有 `DensityDataset` 数据流。

File: electronic_structure/configs/gpwno/README.md
Comment (leeleolay) [inline electronic_structure/configs/gpwno/README.md:1]: log文件是空的

**Comment by leeleolay** [comment]
辛苦补充log文件，保证可自动执行推理命令

File: ppmat/models/gpwno/PWNO_utils.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+                dtype=paddle.complex64,
+            )
+        )
+
+    def compl_mul3d(self, input, weights):
+        return paddle.einsum("bixyz,ioxyz->boxyz", input, weights)
+
+    def forward(self, x):
+        batchsize = x.shape[0]
+        # ✅ 修复1：rfftn的dim传列表（多维FFT要求）
```
Comment (leeleolay) [inline ppmat/models/gpwno/PWNO_utils.py]: 需使用英文注释

File: electronic_structure/configs/gpwno/gpwno_md17_ethane.yaml
Comment (leeleolay) [inline electronic_structure/configs/gpwno/gpwno_md17_ethane.yaml:1]: 参考原论文，补充qm9结果

File: electronic_structure/configs/gpwno/README.md
Comment (leeleolay) [inline electronic_structure/configs/gpwno/README.md:1]: abstract补充插图

**Comment by yqf66** [comment]
1.更新abstract插图
2.补充qm9结果
3.数据集与log文件链接： https://pan.baidu.com/s/1U5QV8phyFtqMcVYq232bvg?pwd=57k2 提取码: 57k2

File: electronic_structure/configs/gpwno/README.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
+# GPWNO
+
+[Gaussian Plane-Wave Neural Operator for Electron Density Estimation](https://arxiv.org/abs/2402.04278)
+
+Official implementation: [seongsukim-ml/GPWNO](https://github.com/seongsukim-ml/GPWNO)
```
Comment (leeleolay) [inline electronic_structure/configs/gpwno/README.md]: 删除

File: electronic_structure/configs/gpwno/README.md [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+\mathrm{NMAE} =
+\frac{\sum_i |\hat{\rho}(x_i)-\rho(x_i)|}
+{\sum_i |\rho(x_i)|}
+$$
+
+---
+
+## Dataset Description
+
+GPWNO uses the same electron-density datasets and dataloaders as InfGCN.
```
Comment (leeleolay) [inline electronic_structure/configs/gpwno/README.md]: 删除这句

File: electronic_structure/configs/gpwno/README.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+- **MD17_EC**: Small-molecule molecular dynamics electron-density data. The supported molecules include benzene, ethanol, phenol, resorcinol, ethane, and malonaldehyde. [Data](https://paddle-org.bj.bcebos.com/paddlematerials/datasets/MD17_ES/md17_es.tar.gz).
+- **QM9_EC**: QM9 molecule electron-density data stored as `qm9.json`, `qm9_data_split.json`, and `*.CHGCAR.lz4` files under `./data/data_qm9`. Data link: TBD.
+- **MP_EC (cubic)**: Materials Project-style crystal electron densities stored as `*.json.xz` under `./data/data_mp`. [Data](https://paddle-org.bj.bcebos.com/paddlematerials/datasets/MP_ES/mp_es.tar), [atom dictionary](https://paddle-org.bj.bcebos.com/paddlematerials/datasets/MP_ES/crystal.json), [split file](https://paddle-org.bj.bcebos.com/paddlematerials/datasets/MP_ES/crystal_data_split.json).
```
Comment (leeleolay) [inline electronic_structure/configs/gpwno/README.md]: 删除，贴数据集自身的描述介绍

**Review by leeleolay — COMMENTED** [review]
复现目标是和原论文的模型指标，训练和推理对应的上

**Comment by leeleolay** [comment]
> 1.更新abstract插图 2.补充qm9结果 3.数据集与log文件链接： https://pan.baidu.com/s/1U5QV8phyFtqMcVYq232bvg?pwd=57k2 提取码: 57k2

辛苦按照已有的格式整理权重文件，应为zip格式，后续会支持一键推理。当前已有infgcn未实现，优化的版本参考https://github.com/PaddlePaddle/PaddleMaterials/pull/302

**Comment by leeleolay** [comment]
未看到md24的相关内容

**Comment by yqf66** [comment]
PR 标题中的 md24 是误标，本次 GPWNO 接入实际覆盖的是原论文使用的 MD、QM9 和 Materials Project（MP）电子密度数据集，并不包含 MD24。现已将 PR 标题修正为 add GPWNO model and MD/QM9/MP configs，同时更新了 README 中的数据集说明、论文指标以及训练配置

File: ppmat/models/__init__.py
Comment (leeleolay) [inline ppmat/models/__init__.py:1]: 未注册模型权重

**Comment by yqf66** [comment]
模型权重压缩包链接：通过网盘分享的文件：gpwno_ckpt.zip
链接: https://pan.baidu.com/s/1o2epYPepzgQ_XBL-G7ywaQ?pwd=2xrf 提取码: 2xrf 
--来自百度网盘超级会员v1的分享

**Comment by leeleolay** [comment]
重构了部分基础组件，移动了推理器的位置，辛苦基于新的开发和尝试 @yqf66

File: ppmat/models/gpwno/GPWNO_utils.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle
+
+# from torch_scatter import scatter
+from paddle_scatter.scatter import scatter
```
Comment (leeleolay) [inline ppmat/models/gpwno/GPWNO_utils.py]: 最新版本已经不依赖paddle_scatter库，辛苦调用ppmat/utils/scatter

File: ppmat/models/__init__.py
Comment (leeleolay) [reply ppmat/models/__init__.py:1]: @yqf66

File: electronic_structure/configs/gpwno/README.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+## Key Configuration
+
+The PaddleMaterials configs follow the original GPWNO hyper-parameter choices:
+
+| Dataset | Original model profile | Main differences |
+| --- | --- | --- |
+| MD17_EC | `GPWNO_MD.yaml` | `num_spherical=3`, `num_fourier=40`, `padding=4`, `use_max_cell=true`, `equivariant_frame=true`, `residual=false` |
+| QM9_EC | `GPWNO_QM9.yaml` | `num_spherical=4`, `num_fourier=40`, `padding=4`, `use_max_cell=true`, `equivariant_frame=true`, `residual=true` |
+| MP_EC | `GPWNO_pbc.yaml` | `pbc=true`, `num_fourier=20`, `padding=0`, `use_max_cell=false`, `max_cell_size=324`, `equivariant_frame=false` |
```
Comment (leeleolay) [inline electronic_structure/configs/gpwno/README.md]: 建议删掉，

File: electronic_structure/configs/gpwno/README.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+---
+
+## Results
+
+The paper reports test NMAE of **2.45%** for MD benzene, **3.67%** for MD ethane, **0.73%** for QM9, and **4.32%** for cubic MP. These are the reproduction targets. The runs below used shorter or otherwise incomplete training schedules and are retained only as implementation smoke-test results; they do not constitute a reproduction of the paper metrics. New checkpoint and log links should be published only after the corresponding test command reaches the paper target with the aligned step budget.
```
Comment (leeleolay) [inline electronic_structure/configs/gpwno/README.md]: 建议删掉

File: electronic_structure/configs/gpwno/README.md [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    </tbody>
+</table>
+
+Note: The MD17_EC_Benzene smoke-test result is from epoch-10 validation of `gpwno_md17_benzene_t_20260527_182931_s_42`; it is not directly comparable to the paper's test result.
+
+Note: The MD17_EC_Ethane smoke-test result is from the test set evaluation of `gpwno_md17_ethane_t_20260529_204928_s_42`. The final validation NMAE is `4.8302%`, and the final test NMAE is `4.8339%`.
+
+Note: The QM9_EC smoke-test result is from the two-epoch test set evaluation of `gpwno_qm9_t_20260626_164201_s_42`, which resumed from `gpwno_qm9_t_20260625_221247_s_42/checkpoints/latest`. The best validation NMAE is `7.0865%`, and the final test NMAE is `7.4001%`.
+
+Note: The MP_EC smoke-test result is from the test set evaluation of `gpwno_mp_resume_t_20260528_204842_s_42`. The final validation NMAE is `34.8928%`, and the final test NMAE is `37.8910%`.
```
Comment (leeleolay) [inline electronic_structure/configs/gpwno/README.md]: 结果在表格里标记，完整的训推结果

File: electronic_structure/configs/gpwno/README.md [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+The datasets are downloaded automatically by the dataset classes when the configured root directory is missing and `auto_download` is enabled.
+
+```bash
+# MD17_EC will be prepared under ./data/data_md
+python electronic_structure/train.py -c electronic_structure/configs/gpwno/gpwno_md17_ethane.yaml Global.do_train=False Global.do_eval=False Global.do_test=True
+
+# QM9_EC will be prepared under ./data/data_qm9 after the dataset package is downloaded and extracted
+python electronic_structure/train.py -c electronic_structure/configs/gpwno/gpwno_qm9.yaml Global.do_train=False Global.do_eval=False Global.do_test=True
+```
```
Comment (leeleolay) [inline electronic_structure/configs/gpwno/README.md]: 删除

File: ppmat/models/gpwno/GPWNO_utils.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        )
+        self.mul = o3.ElementwiseTensorProduct(inner_out, self.vector_irreps)
+
+    def forward(self, features):
+        scalars = self.act_scalars(features[..., : self.scalar_idx])
+        vectors = features[..., self.scalar_idx :]
+        norm = paddle.sqrt(self.inner_prod(vectors, vectors) + 1e-08)
+        act = self.act_vectors(norm)
+        vectors_out = self.mul(act, vectors)
+        return paddle.cat([scalars, vectors_out], dim=-1)
```
Comment (leeleolay) [inline ppmat/models/gpwno/GPWNO_utils.py]: 激活函数放到ppmat/models/common/activation.py 这里吧，可以先看下是否有可以复用的部分

File: ppmat/models/gpwno/GPWNO_utils.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+
+class GCNLayer(paddle.nn.Module):
```
Comment (leeleolay) [inline ppmat/models/gpwno/GPWNO_utils.py]: layer的部分移动到model文件下吧，放到model class 上方，PWNO_utils.py 里的也类似处理吧

File: ppmat/models/gpwno/GPWNO.py [hunk 12]
```diff
@@ -0,0 +1,970 @@
...
+    """
+    Apply periodic boundary condition to the vector
+    :param vec: original vector of (N, K, 3)
+    :param cell: cell frame of (N, 3, 3)
+    :return: shortest vector of (N, K, 3)
+    """
+    coord = vec @ paddle.linalg.inv(x=cell)
+    coord = coord - paddle.round(coord)
+    pbc_vec = coord @ cell
+    return pbc_vec.detach(), coord.detach()
```
Comment (leeleolay) [inline ppmat/models/gpwno/GPWNO.py]: pbc_vec 的 minimum-image 逻辑与 ppmat/models/infgcn/infgcn.py:501-507 中的实现重复。当前仓库虽然没有现成的 pbc_vec helper，但建议将该逻辑抽取到 ppmat/utils/crystal.py，然后由 GPWNO 和 InfGCN 共同复用。建议不要在公共函数中 detach()，以保留对坐标和晶胞的梯度。

File: ppmat/models/common/activation.py
Comment (leeleolay) [inline ppmat/models/common/activation.py:1]: 这个文件的改动？

File: ppmat/models/gpwno/GPWNO.py [hunk 13]
```diff
@@ -0,0 +1,1235 @@
...
+                                dist_min
+                                / self.atom_radius[
+                                    atom_types[batch_idx == b_idx]
+                                ].unsqueeze(0)
+                            ).unsqueeze(-1)
+                        ),
+                        dim=-2,
+                    )
+                minimal_dist_probe[b_idx] = paddle.compat.min(dist_min, dim=-1)[0]
+        if self.input_infgcn:
```
Comment (leeleolay) [inline ppmat/models/gpwno/GPWNO.py:1004]: 这里为什么有个self.input_infgcn的判断

File: ppmat/models/gpwno/GPWNO.py [hunk 14]
```diff
@@ -0,0 +1,1235 @@
...
+        )
+        gridy = paddle.linspace(0, 1, size_y)
+        gridy = gridy.reshape(1, 1, size_y, 1, 1).repeat(
+            [batchsize, size_x, 1, size_z, 1]
+        )
+        gridz = paddle.linspace(0, 1, size_z)
+        gridz = gridz.reshape(1, 1, 1, size_z, 1).repeat(
+            [batchsize, size_x, size_y, 1, 1]
+        )
+        return paddle.cat((gridx, gridy, gridz), dim=-1).to(device)
```
Comment (leeleolay) [inline ppmat/models/gpwno/GPWNO.py:1235]: 这里看起来是构建grid，是否可以复用build_grid

File: ppmat/models/gpwno/GPWNO.py (hunk 13 above)
Comment (yqf66) [reply ppmat/models/gpwno/GPWNO.py:1005]: 这里的 `input_infgcn` 是一个可选的局部密度输入开关：开启后会在 Fourier probe 网格上计算高斯轨道分支的密度，并将其作为额外通道拼接到平面波网络输入。它与 `model_sharing` 不同，后者控制原子特征网络是否共享。

File: ppmat/models/gpwno/GPWNO.py (hunk 14 above)
Comment (yqf66) [reply ppmat/models/gpwno/GPWNO.py]: 这里构建的是 Fourier 网络使用的归一化坐标特征，形状为 `[B, X, Y, Z, 3]`，每个轴通过 `linspace(0, 1, N)` 生成，包含端点 1。

可以按建议复用 `BuildGrid(format="array")`：显式设置原点、网格形状和步长，再将网格坐标转换为对应 dtype/device 的 Paddle Tensor 并扩展 batch。迁移时需要保持原来的端点约定和坐标排列（N>1 时步长为 1/(N-1)，单点轴单独处理），避免改变预训练模型的输入。计划在调整后补充坐标一致性及模型前向验证；目前该调整尚未提交。

Comment (yqf66) [reply ppmat/models/gpwno/GPWNO.py]: 可以复用，现已修改为调用 `BuildGrid(format="array")`，删除原先手动拼接网格的实现。保留了归一化坐标的端点、坐标排列、batch 形状及 dtype/device；单点轴也做了兼容处理。

File: ppmat/models/common/activation.py
Comment (yqf66) [reply ppmat/models/common/activation.py:1]: 这个文件没有修改激活函数的计算逻辑。按前一轮建议，GPWNO 已删除重复实现并复用这里的 `ScalarActivation` / `NormActivation`；检查时发现公共 `ScalarActivation` 内使用了 `Activation` 和 `Extract`，但文件原先未导入这两个类，因此补上了对应 import，避免实例化时出现 NameError。其余差异为空行和文件末尾换行的格式整理。

File: ppmat/models/infgcn/infgcn.py
Comment (leeleolay) [inline ppmat/models/infgcn/infgcn.py:1]: 可以从最新的develop拉取最新的代码，该文件辛苦检查下
